autoagents/agents/agents/search_v3.py

```diff
@@ -0,0 +1,203 @@
+    thought: str
+    reasoning: str
+    plan: List[str]
+    action: str
+    action_input: str
+    citations: List[str]
+    class Config:
+        extra = Extra.forbid
+
+
+def check_valid(o):
+    try:
+        if o.get("action") == "Tool_Finish":
+            FinalOutputSchema(**o)
+        else:
+            InterOutputSchema(**o)
+    except ValidationError:
+        return False
+    return True
+
+
+# Set up a prompt template
+class CustomPromptTemplate(StringPromptTemplate):
+    # The list of tools available
+    tools: List[Tool]
+    ialogger: InteractionsLogger
+
+    def format(self, **kwargs) -> str:
+        # Get the intermediate steps [(AgentAction, Observation)]
+        # Format them in a particular way
+        intermediate_steps = kwargs.pop("intermediate_steps")
+        history = []
+        # Set the agent_scratchpad variable to that value
+        for i, (action, observation) in enumerate(intermediate_steps):
+            if action.tool not in [tool.name for tool in self.tools]:
+                raise Exception("Invalid tool requested by the model.")
+            parsed = json.loads(action.log)
+            if i == len(intermediate_steps) - 1:
+                # Add observation only for the last action
+                parsed["observation"] = observation
+            history.append(parsed)
+        self.ialogger.add_history(history)
+        kwargs["agent_scratchpad"] = json.dumps(history)
+        # Create a tools variable from the list of tools provided
+        kwargs["tools"] = "\n".join([f"{tool.name}: {tool.description}" for tool in self.tools])
+        # Create a list of tool names for the tools provided
+        kwargs["tool_names"] = ", ".join([tool.name for tool in self.tools])
+        kwargs["today"] = date.today()
+        list_prompt =[]
+        list_prompt.append({"role": "goal", "content": kwargs["input"]})
+        list_prompt.append({"role": "tools", "content": [{tool.name: tool.description} for tool in self.tools]})
+        list_prompt.append({"role": "history", "content": history})
+        return json.dumps(list_prompt)
+
+
+class CustomOutputParser(AgentOutputParser):
+    class Config:
+        arbitrary_types_allowed = True
+    ialogger: InteractionsLogger
+    llm: BaseLanguageModel
+    new_action_input: Optional[str]
+    action_history = defaultdict(set)
+
+    def parse(self, llm_output: str) -> Union[AgentAction, AgentFinish]:
+        self.ialogger.add_ai(llm_output)
+        parsed = json.loads(llm_output)
+        if not check_valid(parsed):
+            raise ValueError(f"Could not parse LLM output: `{llm_output}`")
+
+        # Parse out the action and action input
+        action = parsed["action"]
+        action_input = parsed["action_input"]
+
+        if action == "Tool_Finish":
+            return AgentFinish(return_values={"output": action_input}, log=llm_output)
+        return AgentAction(tool=action, tool_input=action_input, log=llm_output)
+
+
+class ActionRunner:
```

Comment:
Can we inherit this from the v2 ActionRunner?

Reply:
We could but at this point v3 ActionRunner has almost all it's subcomponents (which are also interdependent) different from v2 (like tools, prompt, output parser, callback handler etc.) that inheriting from v2 complicates the things, so created a new `ActionRunnerV3`